File: okf/src/okf/bundle/synthesizer.py
from __future__ import annotations

import logging

log = logging.getLogger(__name__)


def _fallback(children: list[tuple[str, str]]) -> str:
    titles = ", ".join(t for t, _ in children[:5] if t) or "concepts"
    more = f" and {len(children) - 5} more" if len(children) > 5 else ""
    return f"Contains {len(children)} concepts ({titles}{more})."


def synthesize_description(
    rel_path: str,
    children: list[tuple[str, str]],
    *,
    model: str = "default",
) -> str:
    if not children:
        return ""
    # Deterministic, lightweight, vendor-neutral directory description synthesis
    descs = [d for _, d in children if d]
    if descs:
        # Use first child's summary pattern or combined count
        return f"Directory containing {len(children)} items, including: {children[0][0]}."
    return _fallback(children)

File: okf/synthesizer.go
package okf

import (
	"fmt"
	"strings"
)

// ChildSummary is a (title, description) pair for a directory entry, used
// as input to description synthesis when generating index.md files.
type ChildSummary struct {
	Title       string
	Description string
}

// SynthesizeDescription produces a deterministic, vendor-neutral summary of
// a directory's contents from its child concepts. relPath is accepted for
// forward compatibility with model-backed synthesis strategies but is not
// currently used.
func SynthesizeDescription(relPath string, children []ChildSummary) string {
	if len(children) == 0 {
		return ""
	}

	for _, c := range children {
		if c.Description != "" {
			return fmt.Sprintf("Directory containing %d items, including: %s.", len(children), children[0].Title)
		}
	}
	return fallbackDescription(children)
}

func fallbackDescription(children []ChildSummary) string {
	n := len(children)
	limit := min(n, 5)

	var titles []string
	for _, c := range children[:limit] {
		if c.Title != "" {
			titles = append(titles, c.Title)
		}
	}
	titleStr := strings.Join(titles, ", ")
	if titleStr == "" {
		titleStr = "concepts"
	}

	more := ""
	if n > 5 {
		more = fmt.Sprintf(" and %d more", n-5)
	}
	return fmt.Sprintf("Contains %d concepts (%s%s).", n, titleStr, more)
}
